lib/services/role.go

```diff
@@ -995,28 +995,77 @@ func (set RoleSet) EnumerateDatabaseUsers(database types.Database, extraUsers ..
 	return result
 }
 
-// EnumerateServerLogins works on a given role set to return a minimal description of allowed set of logins.
-// The wildcard selector is ignored, since it is now allowed for server logins
-func (set RoleSet) EnumerateServerLogins(server types.Server) EnumerationResult {
-	result := NewEnumerationResult()
+// getAllAllowedServerLogins gets all the allowed server
+// logins for the RoleSet.
+func (set RoleSet) getAllAllowedServerLogins() []string {
+	mapped := make(map[string]bool)
+	for _, role := range set {
+		for _, login := range role.GetLogins(types.Allow) {
+			mapped[login] = true
+		}
+		for _, login := range role.GetLogins(types.Deny) {
+			mapped[login] = false
+		}
+	}
+
+	var filtered []string
+	for login, isAllowed := range mapped {
+		if isAllowed {
+			filtered = append(filtered, login)
+		}
+	}
 
-	// gather logins for checking from the roles
-	// no need to check for wildcards
-	var logins []string
+	return filtered
+}
+
+// getAllAllowedServerLogins gets all the allowed server
+// logins for the RoleSet.
+func (set RoleSet) getAllAllowedWindowsDesktopLogins() []string {
```

Comment:
I don't love that the logic is duplicated here, but I don't really care to see it addressed.

You might consider passing in the allow/deny lists, or a func reference that gets them.

Reply:
I did it like this (5fc5d4eed9cd080dcb8f6678c901156956d4599d), however in creating this I think this implementation is wrong.

Currently in `mustGetAllAllowedLogins` I'm grabbing all of the allowed and denied logins in the `RoleSet`, and filtering the denied from the allowed. Then I'm passing that filtered list into `filterAllowedLoginsByResource`, which is filtering out any of those deny-filtered logins that don't have a resource that they apply to. In other words, I'm filtering out all logins a RoleSet denies, regardless of whether that denial applies to the specific desktop/server in question.

For example, consider the case where I have a user with the following two roles in their `RoleSet`:

1. A role which denies `might_exist` logins on a label that `does_not_exist_on_any_desktop`
```yaml
kind: role
metadata:
  name: desktop_label_dne
spec:
  deny:
    windows_desktop_labels:
      label: does_not_exist_on_any_desktop
    windows_desktop_logins:
    - might_exist
version: v5
```

2. A role which allows `might_exist` logins on a label that `does_exist_on_a_desktop`
```yaml
kind: role
metadata:
  name: desktop_label_does_exist
spec:
  allow:
    windows_desktop_labels:
      label: does_exist_on_a_desktop
    windows_desktop_logins:
    - might_exist
version: v5
```

As I've currently implemented things, `might_exist` will _never_ show up in a call to `GetAllowedWindowsDesktopLogins`, because it will get filtered out at the `mustGetAllAllowedLogins` stage. But in this case, I think we want it to show up in the case where the windows desktop passed to `GetAllowedWindowsDesktopLogins` has the `label: does_exist_on_a_desktop` label.

Reply:
It turns out that this is the expected behavior. There's no point in matching `windows_desktop_logins` only to their corresponding labels in a `deny` block, because the label itself means that the entire node is denied -- denying a specific login is meaningless.